zson/builder.go

```diff
@@ -0,0 +1,241 @@
+// into a zng.Value.
+type Builder struct {
+	*zcode.Builder
+}
+
+func NewBuilder() *Builder {
+	return &Builder{
+		Builder: zcode.NewBuilder(),
+	}
+}
+
+func (b *Builder) Build(val Value) (zng.Value, error) {
+	b.Reset()
+	if err := b.buildValue(val); err != nil {
+		return zng.Value{}, err
+	}
+	it := b.Bytes().Iter()
+	bytes, _, err := it.Next()
+	if err != nil {
+		return zng.Value{}, err
+	}
+	return zng.Value{val.TypeOf(), bytes}, nil
+}
+
+func (b *Builder) buildValue(val Value) error {
+	switch val := val.(type) {
+	case *Primitive:
+		return b.buildPrimitive(val)
+	case *Record:
+		return b.buildRecord(val)
+	case *Array:
+		return b.buildArray(val)
+	case *Set:
+		return b.buildSet(val)
+	case *Union:
+		return b.buildUnion(val)
+	case *Map:
+		return b.buildMap(val)
+	case *Enum:
+		return b.buildEnum(val)
+	case *TypeValue:
+		return b.buildTypeValue(val)
+	case *Null:
+		b.AppendNull()
+		return nil
+	}
+	return fmt.Errorf("unknown ast type: %T", val)
+}
+
+func (b *Builder) buildPrimitive(val *Primitive) error {
+	switch zng.AliasedType(val.Type).(type) {
+	case *zng.TypeOfUint8, *zng.TypeOfUint16, *zng.TypeOfUint32, *zng.TypeOfUint64:
+		v, err := strconv.ParseUint(val.Text, 10, 64)
+		if err != nil {
+			return fmt.Errorf("invalid unsigned integer: %s", val.Text)
+		}
+		b.AppendPrimitive(zng.EncodeUint(v))
+		return nil
+	case *zng.TypeOfInt8, *zng.TypeOfInt16, *zng.TypeOfInt32, *zng.TypeOfInt64:
+		v, err := strconv.ParseInt(val.Text, 10, 64)
+		if err != nil {
+			return fmt.Errorf("invalid integer: %s", val.Text)
+		}
+		b.AppendPrimitive(zng.EncodeInt(v))
+		return nil
+	case *zng.TypeOfDuration:
+		v, err := time.ParseDuration(val.Text)
+		if err != nil {
+			return fmt.Errorf("invalid duration: %s", val.Text)
+		}
+		b.AppendPrimitive(zng.EncodeDuration(int64(v)))
+		return nil
+	case *zng.TypeOfTime:
+		t, err := time.Parse(time.RFC3339Nano, val.Text)
+		if err != nil {
+			return fmt.Errorf("invalid ISO time: %s", val.Text)
+		}
+		b.AppendPrimitive(zng.EncodeTime(nano.TimeToTs(t)))
+		return nil
+	case *zng.TypeOfFloat64:
+		v, err := strconv.ParseFloat(val.Text, 64)
+		if err != nil {
+			return fmt.Errorf("invalid floating point: %s", val.Text)
+		}
+		b.AppendPrimitive(zng.EncodeFloat64(v))
+		return nil
+	case *zng.TypeOfBool:
+		var v bool
+		if val.Text == "true" {
+			v = true
+		} else if val.Text != "false" {
+			return fmt.Errorf("invalid bool: %s", val.Text)
+		}
+		b.AppendPrimitive(zng.EncodeBool(v))
+		return nil
+	case *zng.TypeOfBytes:
+		s := val.Text
+		if len(s) < 4 || s[0:2] != "0x" {
+			return fmt.Errorf("invalid bytes: %s", s)
+		}
+		bytes, err := hex.DecodeString(s[2:])
+		if err != nil {
+			return fmt.Errorf("invalid bytes: %s (%s)", s, err.Error())
+		}
+		b.AppendPrimitive(zcode.Bytes(bytes))
+		return nil
+	case *zng.TypeOfString, *zng.TypeOfError:
+		b.AppendPrimitive(zng.EncodeString(val.Text))
+		return nil
+	case *zng.TypeOfBstring:
+		b.AppendPrimitive(unescapeHex([]byte(val.Text)))
```

Comment:
Maybe return an error if val.Text contains a malformed \ sequence?

Reply:
This is checked in the analyzer.